File: Algorithm/129. Sum Root to Leaf Numbers.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-08-04 16:32:06
# @Last modified by:   WuLC
# @Last Modified time: 2016-08-04 16:32:24
# @Email: liangchaowu5@gmail.com

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def sumNumbers(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        result = [0]
        if root!=None:
            self.helper(root, str(root.val), result)
        return result[0]
        
    def helper(self, root, tmp, result):
        if root.left == None and root.right == None:
            result[0] += int(tmp)
            return
        if root.left:
            self.helper(root.left, tmp+str(root.left.val), result)
        if root.right:
            self.helper(root.right, tmp+str(root.right.val), result)
        
        

File: Algorithm/C++/129.sum-root-to-leaf-numbers.cpp
/*
 * @lc app=leetcode id=129 lang=cpp
 *
 * [129] Sum Root to Leaf Numbers
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include <vector>

class Solution {
public:
    int sumNumbers(TreeNode* root) {
        result = 0;
        candidate.clear();
        dfs(root);
        return result;
    }

    void dfs(TreeNode* node) {
        if (node == nullptr) {
            return;
        }
        candidate.push_back(node->val);
        if (node->left == nullptr && node->right == nullptr) {
            int tmp = 0;
            for (int i = 0; i < candidate.size(); i++) {
                tmp = tmp*10 + candidate[i];
            }
            result += tmp;
        } else {
            dfs(node->left);
            dfs(node->right);
        }
        candidate.pop_back();
    }

private:
    int result;
    std::vector<int> candidate;
};

// @lc code=end

